File: python/findClosestValueInBST.py

def findClosestValueInBST( tree, target ):
   return findClosestValueInBSTHelper(tree, target, float("inf"))

def findClosestValueInBSTHelper( tree, target, closest):
  currentNode = tree
  while currentNode is not None:
    if abs(target - closest) > abs(target - currentNode.value):
      closest = currentNode.value
    if target < currentNode.value:
      currentNode = currentNode.left
    elif target > currentNode.value:
      currentNode = currentNode.right
    else:
      break
 
  return closest


# Binary Search Tree
class BST:
  def __init__(self, value):
    self.value = value
    self.left = None
    self.right = None

  def insert(self, value):
    if value < self.value:
      if self.left is None:
        self.left = BST(value)
      else:
        self.left.insert(value)
    else:
      if self.right is None:
        self.right = BST(value)
      else:
        self.right.insert(value)
    return self  

test = BST(10).insert(17).insert(13)
print(findClosestValueInBST(test, 14))

File: go/findClosestValueInBST.go
package main

import (
  "fmt"
  "math"
)

type bstNode struct {
  value int
  left *bstNode
  right *bstNode
}

func (tree *bstNode) FindClosestValueInBST(target int) int {
  return tree.findClosestValueInBST(target, math.MaxInt32)
}

func (tree *bstNode) findClosestValueInBST(target int, closest int) int {
  currentNode := tree
  for currentNode != nil {
    if math.Abs(float64(target) - float64(closest)) > math.Abs(float64(target) - float64(currentNode.value)){
     closest = currentNode.value
    }
    if target < currentNode.value {
      currentNode = currentNode.left
    } else {
      currentNode = currentNode.right
    }
  }
  return closest
}

func New(value int) *bstNode{
  e := &bstNode{ value: value }
  return e
}

func (tree *bstNode) Insert(value int){
  if value < tree.value {
    if tree.left == nil{
       tree.left = &bstNode{value: value}
    } else {
       tree.left.Insert(value)
    }
  } else {
    if tree.right == nil{
      tree.right = &bstNode{value: value}
    } else {
      tree.right.Insert(value)
    }
  }
}

// func main(){
//   bst := New(10)
//   bst.Insert(15)
//   bst.Insert(8)
//   // fmt.Print(bst.value)
//   fmt.Print(bst.FindClosestValueInBST(13))
// }
